### DISCUSSION TITLE
Kymatio vs papers

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#705] [Q&A] [answered]
Kymatio's implementation appears at odds with those of papers, but papers also seem at odds with each other - I hope to find clarification here. Response appreciated.

@janden @lostanlen 

<hr>

1. [JTFS](https://www.lostanlen.com/wp-content/uploads/2019/12/anden2019tsp.pdf) outlines the time scattering network, Fig 5 excerpt below. For first-order, code terminates at `x -> |g_k| -> h^(t)`, but the diagram adds an extra `-> h^(t)` (red)

<img src="https://user-images.githubusercontent.com/16495490/112739340-b4ff0900-8f84-11eb-8912-af3a99bc4646.png" width="400">

2. Code's also missing `x -> |g_k| -> h^(t) -> |g^(t)|` coefficients (orange)
3. JTFS reads as below, which isn't what Kymatio does, and which is rather strange since one bandpass filter cannot cover the full required spectrum. Kymatio properly follows through the continuous-time description with `Q=1`.
   - > To implement the second-order wavelets `psi_mu^(t)`, we use ... a **single bandpass filter** `g^(t)[n]`
4. `psi2` wavelets are [below]; is this explained anywhere? Papers only mention setting `Q=1`.
  a) non-analytic
  b) bimodal; below, red line separates + and - freqs, and the wavelet wraps in with tail from right, and continues to grow toward a second peak in negative freqs. This isn't how Morlet behaves.
  c) pre "periodization", sampled with negative frequencies left, positives right, and the two not splitting at halfway point: `freqs = np.arange((1 - P) * N, P * N)`, breaking correspondence with DFT. What even is "periodization"? It yields the tails, for one.
    - Tails in directly sampled Morlet correspond to high time but low joint resolution, which is unfitting as `|x * psi1|` modulations aren't expected to vary significantly over short spans.

<img src="https://user-images.githubusercontent.com/16495490/112738742-36ec3380-8f7f-11eb-82bb-b3bbb406a82c.png" width="400">

Further on 1 and 2, @janden 's and Mallat's [Deep Scattering Spectrum](https://arxiv.org/pdf/1304.6763.pdf) outlines below diagram, which is consistent with JTFS's continuous-time descriptions, but not discrete diagram. Here, for second-order time scattering, we do the green step above, but not orange or red.

<img src="https://user-images.githubusercontent.com/16495490/112739107-7cf6c680-8f82-11eb-9c2d-fe71c4c577f2.png" width="550">

**Comment by lostanlen** [comment, marked as answer]
Thank you for your questions. 

Questions 1, 2, 3: my IEEE TSP paper with Andén and Mallat proposes to discretize the scattering transform (both TS and jTFS) in terms of a multiresolution pyramid scheme, with several stages of dyadic subsampling.
For a 2-D equivalent, read Sifre and Mallat: https://arxiv.org/pdf/1403.1687.pdf
For an implementation, see: https://github.com/scatnet/scatnet/blob/master/core/wavelet_2d_pyramid.m

Meanwhile, the Kymatio implementation relies on FFTs.
For a description, read chapter 3 of my PhD dissertation: https://tel.archives-ouvertes.fr/tel-01559667/document
For an implementation, see: https://github.com/lostanlen/scattering.m/blob/master/lib/forward/U_to_Y.m

Question 4. This is explained, inter alia:
here https://arxiv.org/abs/1906.09334
here https://arxiv.org/abs/2007.10926
and for Gammatones, in the supplementary material of https://link.springer.com/article/10.1186/s13636-018-0138-4#Sec23
this supplementary material is taken from the Appendix of my dissertation

"Further on 1 and 2 ..." the diagram is correct. It seems like you're conflating the psi/phi pair with the g/h pair.

**Reply by edouardoyallon** [reply]
> Questions 1, 2, 3: my IEEE TSP paper with Andén and Mallat proposes to discretize the scattering transform (both TS and jTFS) in terms of a multiresolution pyramid scheme, with several stages of dyadic subsampling.
> For a 2-D equivalent, read Sifre and Mallat: https://arxiv.org/pdf/1403.1687.pdf
> For an implementation, see: https://github.com/scatnet/scatnet/blob/master/core/wavelet_2d_pyramid.m

Just to  clarify this point, this type of implementation shouldn't be used in practice because the resulting operations are not even close to the filtering with the wavelets (due to aliasing and the fact our wavelets have no compact support) Any implementation with cascade of filter banks should be thought carefully, and in particular not be based on morlet wavelets. Note this paper isn't published.

**Reply by OverLordGoldDragon** [reply]
So Kymatio _isn't_ based on JTFS paper's cascade structure, instead directly discretizing the bottom diagram - makes lot more sense. I'd explicitly note this in JTFS's docs once done. Thanks for your response.

**Reply by lostanlen** [reply]
That's right. The implementation we follow is closer to the MLSP 2015 paper
"Joint time-frequency scattering for audio classification"
https://www.lostanlen.com/wp-content/uploads/2019/06/anden2015mlsp.pdf

than to the IEEE TSP paper you've read. I hope it clarifies the situation

**Reply by OverLordGoldDragon** [reply]
@edouardoyallon I can imagine the cascade handling octaves separably, first with bandpasses on upper frequencies, then lowpass + downsampling on lower (effectively decimation), then repeating bandpass. This seems to account for aliasing and compactness (Kymatio also uses a "tolerance", so not fully devoid of).  Kymatio's approach is more intuitive and performant, but I don't really follow why cascading wouldn't work.

**Reply by edouardoyallon** [reply]
you can't write a gaussian as a product of conjugate mirror filters

**Reply by OverLordGoldDragon** [reply]
@lostanlen Skimming [2](https://arxiv.org/pdf/1906.09334.pdf) and fully reading sections 1 & 2 of [1](https://arxiv.org/pdf/1906.09334.pdf), I've not found an explanation to why `psi2` (Fig 2c) are non-analytic (question 4) - they're not real Morlets either, they seem to be something else entirely. Is this explicitly explained anywhere? Either way seems docs could use correction [here](https://www.kymat.io/codereference.html) in calling `psi2` analytic (and probably Morlet elsewhere).

**Reply by OverLordGoldDragon** [reply]
[This](https://github.com/kymatio/kymatio/blob/master/tests/scattering1d/test_filters_scattering1d.py#L33) explains "periodization", interesting. But still unsure on `psi2` filters; the blue (other post) ones get skipped, and remnants are approximately analytic, but that doesn't seem the goal - `morlet_1d` samples strangely with `freqs` then periodizes.
